recipes/LibriTTS/TTS/tokotron/train.py

```diff
@@ -0,0 +1,989 @@
+            self.hparams.train_logger.log_stats(
+                stats_meta={"epoch": epoch, "lr": lr},
+                train_stats=self.train_stats,
+                valid_stats=stage_stats,
+            )
+
+            # Save the current checkpoint and delete previous checkpoints.
+            self.checkpointer.save_and_keep_only(
+                meta={"loss": stage_stats["loss"]},
+                min_keys=["loss"],
+            )
+
+    def get_bos_prefix(self):
+        """Computes the beginning-of-sequence prefix"""
+        audio_bos_prefix = (
+            torch.ones(
+                self.hparams.bos_width,
+                self.hparams.audio_tokens_per_step,
+                device=self.device,
+            )
+            * hparams["bos_index"]
+        )
+        if self.representation_mode == RepresentationMode.CONTINUOUS:
+            audio_bos_prefix = audio_bos_prefix.unsqueeze(-1).repeat(
+                1, 1, self.hparams.audio_dim
+            )
+        return audio_bos_prefix
+
+    @torch.no_grad()
+    def create_samples(self, batch):
+        """Creates samples for the specified batch
+
+        Arguments
+        ---------
+        batch : PaddedBatch
+            a batch
+        """
+        epoch = self.hparams.epoch_counter.current
+        batch = batch.to(self.device)
+        tokens, tokens_length = batch.tokens
+        infer_out = self.modules.model.infer(
+            input_tokens=tokens, input_length=tokens_length
+        )
+        samples_folder = Path(self.hparams.progress_folder) / str(epoch)
+        samples_folder.mkdir(parents=True, exist_ok=True)
+        wav_length_abs = (infer_out.wav_length * infer_out.wav.size(-1)).int()
+        for uttid, wav, wav_length in zip(
+            batch.uttid, infer_out.wav, wav_length_abs
+        ):
+            wav = wav[:, : wav_length.item()]
+            file_name = samples_folder / f"{uttid}.wav"
+            write_audio(
+                file_name,
+                wav.squeeze(0).detach().cpu(),
+                samplerate=self.hparams.model_sample_rate,
+            )
+
+    def fit_batch(self, batch):
+        """Fit one batch, override to do multiple updates.
+
+        The default implementation depends on a few methods being defined
+        with a particular behavior:
+
+        * ``compute_forward()``
+        * ``compute_objectives()``
+        * ``optimizers_step()``
+
+        Also depends on having optimizers passed at initialization.
+
+        Arguments
+        ---------
+        batch : list of torch.Tensors
+            Batch of data to use for training. Default implementation assumes
+            this batch has two elements: inputs and targets.
+
+        Returns
+        -------
+        loss : torch.Tensor
+            detached loss
+        """
+        loss = super().fit_batch(batch)
+        if self.hparams.lr_annealing_mode == "step":
+            self.hparams.lr_annealing(self.optimizer)
+        return loss
```

Comment:
Maybe use this?
https://speechbrain.readthedocs.io/en/latest/API/speechbrain.core.html#speechbrain.core.Brain.on_fit_batch_end